### PR TITLE
Fix P1 and P2 functional tests 

### PR DESCRIPTION
**Pull request by joelverhagen**
1. Retry calls to `FindPackagesById` since it delegates to search service.
1. Delete tests than take > 10 minutes. Two of the three of that type were already skipped.
1. Fix statistics page web UI test to not depend on a specific package ID.

After this, P1 and P2 tests work just fine against DEV, INT, and PROD.

The whole suite takes 30 - 40 minutes to run.

### CODE REVIEW

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs [hunk 1]
```diff
@@ -63,50 +63,57 @@ public bool CheckIfPackageExistsInSource(string packageId, string sourceUrl)
...
                 }
-                catch (Exception ex)
-                {
-                    WriteLine("Exception thrown while checking the existence of package {0} with version {1}:\r\n {2}", packageId, version, ex.Message);
-                }
+            }
+            catch (Exception ex)
+            {
+                WriteLine($"Exception thrown while executing '{actionName}'.{Environment.NewLine}{ex}");
+                return false;
```
Comment (shishirx34) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: return is unnecessary, return below will return false, wouldn't it?

Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: True, but I wanted to me more explicit. I can remove it.

Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: Fixed.

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs [hunk 2]
```diff
@@ -63,50 +63,57 @@ public bool CheckIfPackageExistsInSource(string packageId, string sourceUrl)
...
+                    {
+                        WriteLine($"[verification attempt {i}]: Waiting {intervalSec} seconds before next check...");
+                        Thread.Sleep(intervalSec * 1000);
+                    }
+
+                    WriteLine($"[verification attempt {i}]: Executing... ");
+                    success = action();
 
-                    for (var i = 0; ((i < maxAttempts) && (!found)); i++)
+                    if (success)
```
Comment (scottbommarito) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: `WriteLine(success ? "Successful!" : "NOT successful!");`

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs [hunk 3]
```diff
@@ -63,50 +63,57 @@ public bool CheckIfPackageExistsInSource(string packageId, string sourceUrl)
...
+        {
+            bool success = false;
             const int intervalSec = 30;
             const int maxAttempts = 30;
 
-            if (success)
+            try
             {
-                try
+                WriteLine($"Starting '{actionName}' ({maxAttempts} attempts, interval {intervalSec} seconds).");
```
Comment (scottbommarito) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: Some extreme nitpicking here, but in order for this to sound 100% correct grammatically, the `actionName` needs to start with a noun. Unfortunately, neither of the two times you call this function do.

Compare:
```
Starting checking if package {packageId} {version} exists on source {sourceUrl}.
```
vs
```
Starting verification that package {packageId} {version} exists on source {sourceUrl}.
```

Completely not necessary, but I would suggest changing the `actionName`s you supply to this function to start with a noun.

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs [hunk 4]
```diff
@@ -203,17 +210,27 @@ public static string GetLatestStableVersion(string packageId)
...
+        public void VerifyVersionCount(string packageId, int expectedCount, bool allowPreRelease = true)
         {
             var repo = PackageRepositoryFactory.Default.CreateRepository(SourceUrl);
-            var packages = repo.FindPackagesById(packageId).ToList();
-            if (!allowPreRelease)
-                packages = packages.Where(item => item.IsReleaseVersion()).ToList();
-            return packages.Count;
+
+            Assert.True(VerifyWithRetry(
+                $"verifying count of {packageId} packages is {expectedCount}",
```
Comment (scottbommarito) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: Another nitpick, but I got confused by the use of `count` here. I would reword the variable names here and wherever else we describe the number of separate version of a package as "count"...
`actualCount` -> `actualNumVersions` etc

File: tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs [hunk 5]
```diff
@@ -154,11 +154,12 @@ private static string GetPackagesAppearInFeedInOrderUrl(DateTime time, string ti
...
...         $@"/GetUpdates()?packageIds='{packageName}%7COwin%7CMicrosoft.Web.Infrastructure%7CMicrosoft.AspNet.Identity.Core%7CMicrosoft.AspNet.Identity.EntityFramework%7CMicrosoft.AspNet.Identity.Owin%7CMicrosoft.AspNet.Web.Optimization%7CRespond%7CWebGrease%7CjQuery%7CjQuery.Validation%7CMicrosoft.Owin.Security.Twitter%7CMicrosoft.Owin.Security.OAuth%7CMicrosoft.Owin.Security.MicrosoftAccount%7CMicrosoft.Owin.Security.Google%7CMicrosoft.Owin.Security.Facebook%7CMicrosoft.Owin.Security.Cookies%7CMicrosoft.Owin%7CMicrosoft.Owin.Host.SystemWeb%7CMicrosoft.Owin.Security%7CModernizr%7CMicrosoft.jQuery.Unobtrusive.Validation%7CMicrosoft.AspNet.WebPages%7CMicrosoft.AspNet.Razor%7Cbootstrap%7CAntlr%7CMicrosoft.AspNet.Mvc%7CNewtonsoft.Json%7CEntityFramework'&versions='" + version1 + "%7C1.0%7C1.0.0.0%7C1.0.0%7C1.0.0%7C1.0.0%7C1.1.1%7C1.2.0%7C1.5.2%7C1.10.2%7C1.11.1%7C2.0.0%7C2.0.0%7C2.0.0%7C2.0.0%7C2.0.0%7C2.0.0%7C2.0.0%7C2.0.0%7C2.0.0%7C2.6.2%7C3.0.0%7C3.0.0%7C3.0.0%7C3.0.0%7C3.4.1.9004%7C5.0.0%7C5.0.6%7C6.0.0'&includePrerelease=false&includeAllVersions=false&targetFrameworks='net45'&versionConstraints='%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C%7C'";
```
Comment (scottbommarito) [inline tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs]: Can we construct in a more formulaic way so it's easier to understand?

```
var packagesList = new List<string> { packageName, "Owin", "Microsoft.Web.Infrastructure", "Microsoft.AspNet.Identity.Core", ... };
var versionsList = new List<string> { version1, "1.0", "1.0.0.0", ... };
var url = UrlHelper.V2FeedRootUrl + 
    @"/GetUpdates()?packageIds='" + 
    string.Join("%7C" /* | */, packagesList) + 
    @"'&versions='" + 
    string.Join("%7C" /* | */, versionsList) + 
    @"'&includePrerelease=false&includeAllVersions=false&targetFrameworks='net45'&versionConstraints='" + 
    (new String("%7C" /* | */, n)) + 
    @"'";
```

Also what is up with the `%7C` repeating at the end? Is it necessary?

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs (hunk 2 above)
Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: sure

File: tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs (hunk 5 above)
Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs]: > Can we construct in a more formulaic way so it's easier to understand?

Sure

> Also what is up with the %7C repeating at the end? Is it necessary?

Part of the original regression test logic. I prefer not to remove that.

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs (hunk 4 above)
Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: Sure

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs (hunk 3 above)
Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: I put it in quotes (to treat it as an abstract phrase) so I didn't have to make the caller think about it. But sure, seems fine

File: tests/NuGetGallery.FunctionalTests/Commandline/PackageVersionTest.cs [hunk 6]
```diff
@@ -22,19 +22,19 @@ public PackageVersionTest(ITestOutputHelper testOutputHelper)
 
         [Fact]
         [Description("Upload multiple versions of a package and see if it gets uploaded properly")]
-        [Priority(0)]
+        [Priority(1)]
         [Category("P1Tests")]
         public async Task UploadMultipleVersionOfPackage()
```
Comment (skofman1) [inline tests/NuGetGallery.FunctionalTests/Commandline/PackageVersionTest.cs]: I wonder if we need this test anymore, since I added a scoped api keys functional test that uploads two packages with same id and different versions.

**Comment by skofman1** [comment]
Looks great! Thanks for making this happend 🌷

File: tests/NuGetGallery.FunctionalTests/Commandline/PackageVersionTest.cs (hunk 6 above)
Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests/Commandline/PackageVersionTest.cs]: That's a great point. I'll take a look today. My guess is that we can axe this!

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs [hunk 7]
```diff
@@ -203,17 +209,27 @@ public static string GetLatestStableVersion(string packageId)
...
-        public int GetVersionCount(string packageId, bool allowPreRelease = true)
+        public void VerifyVersionCount(string packageId, int expectedCount, bool allowPreRelease = true)
         {
             var repo = PackageRepositoryFactory.Default.CreateRepository(SourceUrl);
-            var packages = repo.FindPackagesById(packageId).ToList();
-            if (!allowPreRelease)
-                packages = packages.Where(item => item.IsReleaseVersion()).ToList();
-            return packages.Count;
+
+            Assert.True(VerifyWithRetry(
```
Comment (dtivel) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: Will introducing retry logic here hide failures that we should be investigating?

Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: Great question, Damon. In this case, the retry loop is to account for a well known delay between publish time and availability in the search service. This test assertion used `FindPackagesById` on the gallery's V2 API, which delegates to the search service.

Comment (dtivel) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: It would be beneficial to maintainers of this code if you added a comment to this effect to the `VerifyWithRetry(...)` method.  Otherwise, the _why_ behind retrying isn't clear.

Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: Good idea! I'll do that.

File: tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs (hunk 5 above)
Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs]: This comments lead to a great conversation between Scott and I. We are trying to balance clarity of test and capturing the original user scenario as-is (not a simplification of the scenario which may actually not cover the user's scenario). In the end, I think this is a bit of a moot point since the URL in https://github.com/NuGet/NuGetGallery/issues/1199 does not match the one in the test.

I think I'll take the URL from the original issue and use that one, instead of this long ugly URL.

File: tests/NuGetGallery.FunctionalTests/Commandline/PackageVersionTest.cs (hunk 6 above)
Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests/Commandline/PackageVersionTest.cs]: I think I'm want to keep the test since it uses `FindPackagesById`, which the scoped API key test does not. The scoped API key test already does a lot so keeping them separate makes sense.

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs (hunk 7 above)
Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: Fixed.

File: tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs (hunk 5 above)
Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs]: Fixed.

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs (hunk 4 above)
Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: Fixed.

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs (hunk 2 above)
Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: Fixed.

File: tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs (hunk 3 above)
Comment (joelverhagen) [inline tests/NuGetGallery.FunctionalTests.Core/Helpers/ClientSdkHelper.cs]: Fixed.

**Comment by joelverhagen** [comment]
Addresses https://github.com/NuGet/NuGetGallery/issues/3540.
